### DISCUSSION TITLE
Converting Icecast stream to HLS without re encoding

### DISCUSSION THREAD

**Discussion by Rafi594** [savonet/liquidsoap#3256] [Show and tell]
Hello, 
I made a little script to convert Icecast stream to HLS without re encoding, with the latest rolling release metadata is forwarded too !

The only thing I've not tested is adding multiple streams for different quality but I think it's not possible currently.

Also here I'm using /dev/shm to store the HLS files to avoid reducing SSD life too much but you can use any folder you want to.

```
s = input.http("http://YourIcecastServer:8000/stream-aac")

aac_hifi = %ffmpeg(format="mpegts",
                   %audio.copy)

streams = [("aac_hifi", aac_hifi.{ id3 = true, id3_version = 3 })]

def segment_name(~position,~extname,stream_name) =
  timestamp = int_of_float(time())
  duration = 2
  "#{stream_name}_#{duration}_#{timestamp}_#{position}.#{extname}"
end

output.file.hls(playlist="live.m3u8",
                fallible=true,
                segment_duration=5.0,
                segments=5,
                segments_overhead=5,
                segment_name=segment_name,
                persist_at="station.config",
                "/dev/shm/streams/station",
                streams,
                s)
```

**Comment by toots** [comment]
Looks great thank you for sharing!

**Comment by christian-gama** [comment]
What if I want to have different bitrates (lofi, midfi, hifi) without reencoding? I can make it with FFmpeg, using the -c copy, but in Liquidsoap %audio.copy does not allow to pass any param such as bitrate.

**Reply by toots** [reply]
How would you achieve it with ffmpeg?

**Reply by christian-gama** [reply]
> How would you achieve it with ffmpeg?

I was about to send you the exact FFmpeg I was using, but I just realized it was not changing the bitrate at all. My bad, I thought I could do that. Anyway, the command I was using was something like this:

```sh
ffmpeg -i pipe:0 -c copy -b:a 128 -f mpegts pipe:1
```

But I just realized that the bitrate was not changing at all. As FFmpeg did not warn me, I always thought I was indeed changing the bitrate. To change the bitrate I have to specify the `-c:a` flag.

**Reply by toots** [reply]
Cool yeah that's what I was thinking.

I could still be possible to do in liquidsoap but it'd require a lot of prep on the user side, having all files pre-emptively transcoded to the right bitrate etc.

If you want to file a feature request I'd be happy to take a look if time allows.

**Comment by chodgetts** [comment]
Hello, 
I know this was almost 2 years ago, but I am curious when you say " with the latest rolling release metadata is forwarded too " 

I am not seeing any metadata when running this script? 
I am getting the HLS generated, but from my Icecast which only has StreamTitle included, I am not getting that transferred to the HLS stream, either as an ID3 or in the .m3u8 file as #EXTINF data - as in the .m3u8 file does not contain it. 

So wondering what I might be doing wrong here :(

**Reply by toots** [reply]
Yeah it looks like you're gonna need to use a `metadata.map` operator to split that icecast specific metadata into `title` and `artist`.

**Reply by chodgetts** [reply]
Thank you so much for that.. 
I am struggling with this :( - but this is my first experience with liquidsoap. 
I will get there, but just wanted to thank you for the suggestion :)
